eta/core/serial.py

```diff
@@ -812,6 +818,550 @@ def _validate_dict(cls, d):
+        invalidates the current Big iterable.
+
+        Args:
+            archive_path: the path + extension to write the output archive
+            delete_backing_dir: whether to delete the original backing
+                directory when creating the archive. By default, this is False
+        '''
+        with etau.TempDir() as tmp_dir:
+            name = os.path.splitext(os.path.basename(archive_path))[0]
+            rootdir = os.path.join(tmp_dir, name)
+            index_path = os.path.join(rootdir, "index.json")
+            ele_dir = os.path.join(rootdir, self._ELE_ATTR)
+
+            if delete_backing_dir:
+                self.move(backing_dir=ele_dir)
+                big = self
+            else:
+                big = self.copy(backing_dir=ele_dir)
+
+            #
+            # The backing directory embedded in the JSON is not actually used
+            # (neither here nor in `from_archive`), but we set it relative to
+            # the root of the archive, for completeness.
+            #
+            full_backing_dir = big.backing_dir
+            big._backing_dir = "./" + self._ELE_ATTR
+            big.write_json(index_path)
+            big._backing_dir = full_backing_dir
+
+            etau.make_archive(rootdir, archive_path)
+
+    @classmethod
+    def from_archive(
+            cls, archive_path, backing_dir=None, delete_archive=False):
+        '''Loads a Big iterable from an archive created by `to_archive()`.
+
+        Args:
+            archive_path: the path to the archive to load
+            backing_dir: optional backing directory to use for the new Big
+                iterable. If provided, it must be empty or non-existent
+            delete_archive: whether to delete the archive after unpacking it.
+                By default, this is False
+
+        Returns:
+            a Big iterable
+        '''
+        with etau.TempDir() as tmp_dir:
+            name = os.path.splitext(os.path.basename(archive_path))[0]
+            rootdir = os.path.join(tmp_dir, name)
+            index_path = os.path.join(rootdir, "index.json")
+            tmp_backing_dir = os.path.join(rootdir, cls._ELE_ATTR)
+
+            etau.extract_archive(
+                archive_path, outdir=tmp_dir, delete_archive=delete_archive)
+            big = cls.from_json(index_path)
+            big._backing_dir = tmp_backing_dir
+            big.move(backing_dir=backing_dir)
+
+        return big
+
+    @classmethod
+    def from_paths(cls, paths, backing_dir=None):
+        '''Creates a Big iterable from a list of `_ELE_CLS` JSON files.
+
+        Args:
+            backing_dir: optional backing directory to use for the new Big
+                iterable. If provided, it must be empty or non-existent
+            paths: an iterable of paths to `_ELE_CLS` JSON files
+
+        Returns:
+            a Big iterable
+        '''
+        big = cls(backing_dir=backing_dir)
+        for path in paths:
+            big._add_by_path(path)
```

Comment:
`BigMixin` would need to declare `_add_by_path()` as an abstract method that subclasses need to implement.

Given my comments on the private methods below, how about instead making `add_by_path(path)` a public method of `BigMixin` that all subclasses must implement?

(implementation detail: `BigSet.add_by_path(path, key=None)` can additionally allow callers to provide the `key` to avoid reading the element, which wouldn't be used by `BigMixin.from_paths()`)

Reply:
Addressed